src/System.Device.Gpio/System/Device/Gpio/Drivers/LibgpiodDriver.cs

```diff
@@ -0,0 +1,190 @@
+
+        protected internal override PinMode GetPinMode(int pinNumber)
+        {
+            if (_pinNumperToPinModes.ContainsKey(pinNumber))
+             {
+                return _pinNumperToPinModes[pinNumber];
+            }
+            else
+            {
+                throw new InvalidOperationException("There was an attempt to get a mode to a pin that is not yet open.");
+            }
+        }
+
+        protected internal override bool IsPinModeSupported(int pinNumber, PinMode mode)
+        {
+            // Libgpiod driver does not support pull up or pull down resistors.
+            if (mode == PinMode.InputPullDown || mode == PinMode.InputPullUp)
+                return false;
+            return true;
+        }
+
+        protected internal override void OpenPin(int pinNumber)
+        {
+            // do nothing
+        }
+
+        protected internal override PinValue Read(int pinNumber)
+        {
+            PinValue result = default(PinValue);
+            try
+            {
+                Process myProc = new Process();
+                myProc.StartInfo.FileName = GpioGet;
+                myProc.StartInfo.Arguments = $" {GpioChip0} {pinNumber}";
+                myProc.StartInfo.RedirectStandardOutput = true;
+                myProc.Start();
+                if (HandlProcessExit(myProc, $"{GpioGet} {GpioChip0} {pinNumber}", _waitMilliSeconds))
+                {
+                    string valueContents = myProc.StandardOutput.ReadToEnd();
+                    result = ConvertSysFsValueToPinValue(valueContents);
+                }
+            }
+            catch (UnauthorizedAccessException e)
+            {
+                throw new UnauthorizedAccessException("Reading a pin value requires root permissions.", e);
+            }
+            return result;
+        }
+        private PinValue ConvertSysFsValueToPinValue(string value)
+        {
+            PinValue result;
+            value = value.Trim();
+
+            switch (value)
+            {
+                case "0":
+                    result = PinValue.Low;
+                    break;
+                case "1":
+                    result = PinValue.High;
+                    break;
+                default:
+                    throw new ArgumentException($"Invalid Gpio pin value {value}");
+            }
+
+            return result;
+        }
+
+        protected internal override void RemoveCallbackForPinValueChangedEvent(int pinNumber, PinChangeEventHandler callback)
+        {
+            throw new NotImplementedException();
+        }
+
+        protected internal override void SetPinMode(int pinNumber, PinMode mode)
+        {
+            if (mode == PinMode.InputPullDown || mode == PinMode.InputPullUp)
+            {
+                throw new PlatformNotSupportedException("This driver is generic so it does not support Input Pull Down or Input Pull Up modes.");
+            }
+            _pinNumperToPinModes[pinNumber] = mode;
```

Comment:
Looks like this isn't really setting the mode at all. You are only keeping the Mode state, but not really affecting the real mode in the hardware. I can imagine there will be some scenarios where we only want to change the mode, without having to read or write.

Reply:
Yes it is logical only to suport the API methods, as i mentioned before command line tools doesn't have any options for setting pin mode, getpin, setpin commands does that internally, after getpin command the pin became input, after setpin it became output automatically.

Reply:
I will then call get or set pin functions in order to make sure that the modes on the pins are actually changed after calling this.

Reply:
If the command line tool does not support PullUp or PullDown then we can't use them.  This is a very common scenario that should be supported.

Reply:
I wouldn't spend too much time on this - I presume using Process is only to give us idea if the lib is capable of doing what we want and then eventually we will use the lib directly